### PR TITLE
Simplify db code

### PR DESCRIPTION
**Pull request by Bond-009**
Back to 1 single connection/DB.

### CODE REVIEW

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 1]
```diff
@@ -2842,75 +2817,72 @@ public QueryResult<BaseItem> GetItems(InternalItemsQuery query)
                 statementTexts.Add(commandText);
             }
 
-            using (WriteLock.Read())
+            using (var connection = CreateConnection(true))
             {
-                using (var connection = CreateConnection(true))
+                return connection.RunInTransaction(db =>
```
Comment (cvium) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: It should not be run in a transaction.

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 2]
```diff
@@ -3262,64 +3228,61 @@ public QueryResult<Guid> GetItemIds(InternalItemsQuery query)
...
```
Comment (cvium) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: Same here

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 3]
```diff
@@ -5480,100 +5428,97 @@ private QueryResult<(BaseItem, ItemCounts)> GetItemValues(InternalItemsQuery que
                 statementTexts.Add(countText);
             }
 
-            using (WriteLock.Read())
+            using (var connection = CreateConnection(true))
             {
-                using (var connection = CreateConnection(true))
+                return connection.RunInTransaction(db =>
```
Comment (cvium) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: And here

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 4]
```diff
@@ -2842,75 +2817,68 @@ public QueryResult<BaseItem> GetItems(InternalItemsQuery query)
...
 
-            using (WriteLock.Read())
+            using (var connection = GetConnection(true))
             {
-                using (var connection = CreateConnection(true))
+                var statements = PrepareAllSafe(connection, statementTexts);
+
+                if (!isReturningZeroItems)
                 {
-                    return connection.RunInTransaction(db =>
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: I'm wary that if you run without locking and without transactions, then those two queries could potentially return different results if DB was changed between those...

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 5]
```diff
@@ -3262,64 +3224,58 @@ public QueryResult<Guid> GetItemIds(InternalItemsQuery query)
...
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: same worry here

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 6]
```diff
@@ -5480,100 +5421,95 @@ private QueryResult<(BaseItem, ItemCounts)> GetItemValues(InternalItemsQuery que
...
+            using (var connection = GetConnection(true))
             {
-                using (var connection = CreateConnection(true))
+                var list = new List<(BaseItem, ItemCounts)>();
+
+                var statements = PrepareAllSafe(connection, statementTexts);
+
+                if (!isReturningZeroItems)
                 {
-                    return connection.RunInTransaction(db =>
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: and here...

File: Emby.Server.Implementations/Data/SqliteUserDataRepository.cs [hunk 7]
```diff
@@ -397,10 +380,5 @@ protected override void Dispose(bool dispose)
         {
             // handled by library database
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/SqliteUserDataRepository.cs]: maybe drop this empty override?

**Comment by ploughpuff** [comment]
I'm unsure how this can be working..  Has it been tested?

Allowing on-demand db connections and without the locks on WRITE, you're going to receive a lot of BUSY notifications back from SQLite, which SQLitePCL.pretty handles by throwing an exception.  Not good!

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 8]
```diff
@@ -31,130 +29,96 @@ protected BaseSqliteRepository(ILogger logger)
...
-                    connectionFlags |= ConnectionFlags.ReadWrite;
+                    _defaultWal = db.Query("PRAGMA journal_mode").SelectScalarString().First();
+
+                    Logger.LogInformation("Default journal_mode for {0} is {1}", DbFilePath, _defaultWal);
                 }
-                else
+
+                var queries = new List<string>
+                {
+                    //"PRAGMA cache size=-10000"
```
[the remark this answers is not in the document]
Comment (ploughpuff) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: The negative sign indicates bytes instead of pages.

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs (hunk 4 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/Data/SqliteItemRepository.cs]: More likely, SQLite will return a LOCKED notification back to the reader which will result in an exception..  Using transaction for SELECT (READS) is not required and eventually results in 'transaction in transaction' errors.

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 9]
```diff
@@ -199,7 +163,7 @@ protected bool TableExists(ManagedConnection connection, string name)
             }, ReadTransactionMode);
         }
 
-        protected void RunDefaultInitialization(ManagedConnection db)
+        protected void RunDefaultInitialization(SQLiteDatabaseConnection db)
```
Comment (ploughpuff) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: Can't see where this function is called for your new connections?  Not sure the consequences of missing out these PRAGMAS..

**Comment by joshuaboniface** [comment]
@Bond-009 Where are we at with this one, proceed or close?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 10]
```diff
@@ -27,123 +27,107 @@ protected BaseSqliteRepository(ILogger logger)
...
+        public IStatement PrepareStatementSafe(ManagedConnection connection, string sql)
+            => connection.PrepareStatement(sql);
+
         public IStatement PrepareStatement(IDatabaseConnection connection, string sql)
-        {
-            return connection.PrepareStatement(sql);
-        }
+            => connection.PrepareStatement(sql);
 
         public IStatement PrepareStatementSafe(IDatabaseConnection connection, string sql)
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: why do we need two functions doing effectively the same?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 11]
```diff
@@ -27,123 +27,107 @@ protected BaseSqliteRepository(ILogger logger)
...
+                WriteLock.Wait();
+                return new ManagedConnection(WriteConnection, WriteLock);
+            }
         }
 
+        public IStatement PrepareStatement(ManagedConnection connection, string sql)
+            => connection.PrepareStatement(sql);
+
+        public IStatement PrepareStatementSafe(ManagedConnection connection, string sql)
+            => connection.PrepareStatement(sql);
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: why do we need two functions doing effectively the same?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 12]
```diff
@@ -224,6 +205,25 @@ protected virtual void Dispose(bool dispose)
                 return;
             }
 
+            if (dispose)
+            {
+                WriteLock.Wait();
+                try
+                {
+                    WriteConnection.Dispose();
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: shouldn't you do `WriteConnection = null` here as well?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 13]
```diff
@@ -27,123 +27,107 @@ protected BaseSqliteRepository(ILogger logger)
...
+                if (WriteConnection == null)
+                {
+                    throw new InvalidOperationException("Can't access the write connection at this time.");
+                }
 
-        public IStatement PrepareStatementSafe(SQLiteDatabaseConnection connection, string sql)
-        {
-            return connection.PrepareStatement(sql);
+                WriteLock.Wait();
+                return new ManagedConnection(WriteConnection, WriteLock);
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: why cannot `ManagedConnection` take this lock? it's obviously going to be releasing it, so why it cannot acquire it, too?

File: Emby.Server.Implementations/Data/ManagedConnection.cs [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            _writeLock?.Release();
+            _readConPool?.Add(_db);
+
+            _db = null; // Don't dispose it
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/ManagedConnection.cs]: why?

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 15]
```diff
@@ -614,7 +613,7 @@ private void SaveItemsInTranscation(IDatabaseConnection db, IEnumerable<(BaseIte
             {
                 GetSaveItemCommandText(),
                 "delete from AncestorIds where ItemId=@ItemId"
-            });
+            }).ToList();
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: why suddenly `ToList`?

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 16]
```diff
@@ -2817,12 +2816,12 @@ public QueryResult<BaseItem> GetItems(InternalItemsQuery query)
...
             }
 
-            using (var connection = CreateConnection(true))
+            using (var connection = GetConnection(true))
             {
                 return connection.RunInTransaction(db =>
                 {
                     var result = new QueryResult<BaseItem>();
-                    var statements = PrepareAllSafe(db, statementTexts);
+                    var statements = PrepareAllSafe(db, statementTexts).ToList();
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: why `ToList`?

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 17]
```diff
@@ -3228,13 +3227,13 @@ public QueryResult<Guid> GetItemIds(InternalItemsQuery query)
...
 
-            using (var connection = CreateConnection(true))
+            using (var connection = GetConnection(true))
             {
                 return connection.RunInTransaction(db =>
                 {
                     var result = new QueryResult<Guid>();
 
-                    var statements = PrepareAllSafe(db, statementTexts);
+                    var statements = PrepareAllSafe(db, statementTexts).ToList();
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: again

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 18]
```diff
@@ -5428,14 +5427,14 @@ private QueryResult<(BaseItem, ItemCounts)> GetItemValues(InternalItemsQuery que
...
-            using (var connection = CreateConnection(true))
+            using (var connection = GetConnection(true))
             {
                 return connection.RunInTransaction(db =>
                 {
                     var list = new List<(BaseItem, ItemCounts)>();
                     var result = new QueryResult<(BaseItem, ItemCounts)>();
 
-                    var statements = PrepareAllSafe(db, statementTexts);
+                    var statements = PrepareAllSafe(db, statementTexts).ToList();
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: again

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (JustAMan) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: I was wondering about comments, why are they inside, why don't just remove those?..

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 13 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: IMO that's not something you should do in a constructor.

File: Emby.Server.Implementations/Data/ManagedConnection.cs (hunk 14 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/ManagedConnection.cs]: In the case of readonly connections we returned it to the pool, for write connections we just use the one protected by the lock. We don't want to dispose those.

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs (hunk 15 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/SqliteItemRepository.cs]: Before the function returned a list ( it did `ToList` internally), now it returns an `IEnumerable`

File: Emby.Server.Implementations/Security/AuthenticationRepository.cs [hunk 19]
```diff
@@ -362,31 +351,28 @@ private static AuthenticationInfo Get(IReadOnlyList<IResultSetValue> reader)
 
         public DeviceOptions GetDeviceOptions(string deviceId)
         {
-            using (WriteLock.Read())
+            using (var connection = GetConnection(true))
             {
-                using (var connection = CreateConnection(true))
+                return connection.RunInTransaction(db =>
```
Comment (ploughpuff) [inline Emby.Server.Implementations/Security/AuthenticationRepository.cs]: Given this is a read only query (SELECT), you might have to remove the RunInTransaction to avoid errors like 'cannot start transaction in transaction'.

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 20]
```diff
@@ -29,155 +27,105 @@ protected BaseSqliteRepository(ILogger logger)
...
+            }
+        }
 
-                    foreach (var query in queries)
-                    {
-                        db.Execute(query);
-                    }
-                }
-                catch
+        protected ManagedConnection GetConnection(bool isReadOnly = false)
```
[the remark this answers is not in the document]
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: Why? You have only 2 possible values + I tried to reduce breakage.

**Comment by ploughpuff** [comment]
Hmmm... Brief test of the latest changes and unfortunately I hit an exception pretty quickly:

`Locked: database table is locked SQLitePCL.pretty.SQLiteException: Exception of type 'SQLitePCL.pretty.SQLiteException' was thrown.`

This was thrown from GetItemList() which is using a RO connection.  Need to understand what's going on here as LOCKED indicates a conflict whilst using the same connection, and we've got BlockingCollection (mutex?) protection to guard against that..

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (ploughpuff) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: GetConnection(true) ain't as clear as GetConnection(ReadOnly), that's all.

**Comment by Bond-009** [comment]
Wooops

**Comment by ploughpuff** [comment]
Indeed.  Couple of points:
1) Maybe look to only have shared cache between RO connections, not with WR..
2) Are you now creating two WR connections to library db?

**Comment by Bond-009** [comment]
I mentioned in the dev chat that I’m scratching the multiple connections all together. Perf was gained in onother place and made the added complexity not worth it.

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (JustAMan) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: I second @ploughpuff here. Plus I suggest removal of default value, IMHO for clarity `GetConnection()` should always have an argument

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 21]
```diff
@@ -25,88 +23,45 @@ protected BaseSqliteRepository(ILogger logger)
...
-        {
-            ThreadSafeMode = raw.sqlite3_threadsafe();
-            raw.sqlite3_enable_shared_cache(1);
-        }
+        private SQLiteDatabaseConnection _writeConnection;
 
         private string _defaultWal;
 
-        protected async Task CreateConnections()
+        protected ManagedConnection GetConnection(bool _ = false)
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: why are you leaving `_` parameter here?

Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: Didn’t want to break all uses in this PR.

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 10 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: One is for `IDatabaseConnection` and the other for `ManagedConnection`

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 11 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: above

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 12 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: No, the field is readonly, can't do that and isn't really needed.

File: Emby.Server.Implementations/Data/SqliteUserDataRepository.cs [hunk 22]
```diff
@@ -33,14 +32,9 @@ public class SqliteUserDataRepository : BaseSqliteRepository, IUserDataRepositor
         /// Opens the connection to the database
         /// </summary>
         /// <returns>Task.</returns>
-        public void Initialize(ReaderWriterLockSlim writeLock, ManagedConnection managedConnection, IUserManager userManager)
```
Comment (ploughpuff) [inline Emby.Server.Implementations/Data/SqliteUserDataRepository.cs]: I believe the original approach taken here results in two 'write' connections being created to library.db.  Whilst this is safe in practice (we have mutex protection etc.) it may result in conflicts when both (SqliteItemRepo and SqliteUserDataRepo) want to write at the same time; one shall receive a LOCK error status.

All the other db files only have one connection being made to them.  Could we modify library to be the same?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 23]
```diff
@@ -12,172 +11,76 @@ namespace Emby.Server.Implementations.Data
...
 
-        internal static int ThreadSafeMode { get; set; }
-
-        static BaseSqliteRepository()
-        {
-            SQLite3.EnableSharedCache = false;
-
-            int rc = raw.sqlite3_config(raw.SQLITE_CONFIG_MEMSTATUS, 0);
-            //CheckOk(rc);
+        protected virtual ConnectionFlags DefaultConnectionFlags => ConnectionFlags.SharedCached | ConnectionFlags.NoMutex;
```
Comment (ploughpuff) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: SharedCache necessary now we've only one connection per db?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 11 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: I think the idea is to remove the 'Safe' version as it's identical (same input params too).

File: Emby.Server.Implementations/Data/SqliteUserDataRepository.cs (hunk 22 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/Data/SqliteUserDataRepository.cs]: My comment above is more about the instantiation, creating two db connections:
`public class SqliteUserDataRepository : BaseSqliteRepository, IUserDataRepository`

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 11 above)
Comment (JustAMan) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: yes

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 10 above)
Comment (JustAMan) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: why have `Safe` and non-`Safe` identical functions?

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs (hunk 15 above)
Comment (JustAMan) [reply Emby.Server.Implementations/Data/SqliteItemRepository.cs]: But why you need a list here, why `IEnumerable` isn't enough?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 23 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: Doesn't hurt anything right?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 24]
```diff
@@ -243,7 +207,7 @@ protected void CheckDisposed()
 
         public void Dispose()
         {
-            _disposed = true;
+
             Dispose(true);
```
[the remark this answers is not in the document]
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: Yes this is needed, the deriving class should override the `Dispose(bool)` function.

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 25]
```diff
@@ -255,42 +219,13 @@ public void Dispose()
         /// <param name="dispose"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
         protected virtual void Dispose(bool dispose)
```
[the remark this answers is not in the document]
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: Yes

File: Emby.Server.Implementations/Data/SqliteUserDataRepository.cs (hunk 22 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/SqliteUserDataRepository.cs]: This is... dirty, 2 classes accessing the same DB and sharing a lock.

File: Emby.Server.Implementations/Security/AuthenticationRepository.cs (hunk 19 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Security/AuthenticationRepository.cs]: I'm keeping my hands of these for now.

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 11 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: Done

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 10 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: Removed them

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 23 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: hurt, no, performance, maybe.. but think you've removed it now anyway?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 26]
```diff
@@ -1,183 +1,76 @@
...
-                    var queries = new List<string>
-                    {
-                        //"PRAGMA cache size=-10000"
-                        //"PRAGMA read_uncommitted = true",
-                        "PRAGMA synchronous=Normal"
-                    };
-
-                    if (CacheSize.HasValue)
-                    {
-                        queries.Add("PRAGMA cache_size=" + CacheSize.Value.ToString(CultureInfo.InvariantCulture));
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: why is this pragma gone?

File: Emby.Server.Implementations/Data/ManagedConnection.cs (hunk 14 above)
Comment (JustAMan) [reply Emby.Server.Implementations/Data/ManagedConnection.cs]: Is this still relevant?

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 27]
```diff
@@ -551,21 +549,18 @@ public void SaveImages(BaseItem item)
...
             CheckDisposed();
 
-            using (WriteLock.Write())
+            using (var connection = GetConnection())
             {
-                using (var connection = CreateConnection())
+                connection.RunInTransaction((Action<IDatabaseConnection>)(db =>
                 {
-                    connection.RunInTransaction(db =>
+                    using (var saveImagesStatement = base.PrepareStatement((IDatabaseConnection)db, (string)"Update TypedBaseItems set Images=@Images where guid=@Id"))
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: why explicit cast to `string`?

File: Emby.Server.Implementations/Data/SqliteUserDataRepository.cs (hunk 7 above)
Comment (JustAMan) [reply Emby.Server.Implementations/Data/SqliteUserDataRepository.cs]: bump?

File: Emby.Server.Implementations/Security/AuthenticationRepository.cs [hunk 28]
```diff
@@ -362,31 +346,28 @@ private static AuthenticationInfo Get(IReadOnlyList<IResultSetValue> reader)
 
         public DeviceOptions GetDeviceOptions(string deviceId)
         {
-            using (WriteLock.Read())
+            using (var connection = GetConnection(true))
             {
-                using (var connection = CreateConnection(true))
+                return connection.RunInTransaction((Func<IDatabaseConnection, DeviceOptions>)(db =>
```
Comment (JustAMan) [inline Emby.Server.Implementations/Security/AuthenticationRepository.cs]: why cast?

File: Emby.Server.Implementations/Security/AuthenticationRepository.cs [hunk 29]
```diff
@@ -362,31 +346,28 @@ private static AuthenticationInfo Get(IReadOnlyList<IResultSetValue> reader)
...
                 {
-                    return connection.RunInTransaction(db =>
+                    using (var statement = base.PrepareStatement((IDatabaseConnection)db, (string)"select CustomName from Devices where Id=@DeviceId"))
```
Comment (JustAMan) [inline Emby.Server.Implementations/Security/AuthenticationRepository.cs]: why cast?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 30]
```diff
@@ -1,183 +1,76 @@
...
-        public List<IStatement> PrepareAll(IDatabaseConnection connection, IEnumerable<string> sql)
-        {
-            return PrepareAllSafe(connection, sql);
-        }
-
-        public List<IStatement> PrepareAllSafe(IDatabaseConnection connection, IEnumerable<string> sql)
-        {
-            return sql.Select(connection.PrepareStatement).ToList();
-        }
+        public IEnumerable<IStatement> PrepareAllSafe(IDatabaseConnection connection, IEnumerable<string> sql)
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: why are you keeping `PrepareAllSafe` instead of `PrepareAll`? other functions with `Safe` in them got dropped...

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 26 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: It was always null, it never got executed.
`protected virtual int? CacheSize => null;`

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs (hunk 27 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/SqliteItemRepository.cs]: Weird, I didn't touch this. I'll change it back.

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 26 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: What about the override in SqliteItemRepo:
protected override int? CacheSize => 20000;

File: Emby.Server.Implementations/Data/SqliteUserDataRepository.cs (hunk 22 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/Data/SqliteUserDataRepository.cs]: We either need to raise a ticket so this issue is not forgotten, or add a ToDo (if anyone cares checking those)..

File: Emby.Server.Implementations/Data/SqliteUserRepository.cs [hunk 31]
```diff
@@ -75,10 +75,12 @@ private void TryMigrateToLocalUsersTable(ManagedConnection connection)
             }
         }
 
-        private void RemoveEmptyPasswordHashes()
+        private void RemoveEmptyPasswordHashes(ManagedConnection connection)
         {
-            foreach (var user in RetrieveAllUsers())
+            foreach (var row in connection.Query("select id,guid,data from LocalUsersv2"))
```
Comment (JustAMan) [inline Emby.Server.Implementations/Data/SqliteUserRepository.cs]: why change?

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 26 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: I'll add the logic back for the possibility to easily change this, but I thing we should do some testing to see if it's really benefiting anything. The same goes for `PRAGMA synchronous`

File: Emby.Server.Implementations/Data/SqliteUserRepository.cs (hunk 31 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/SqliteUserRepository.cs]: `RetrieveAllUsers` does another `GetConnection` which will block indefinitely because we already got a connection.

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs (hunk 30 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: Fixed

File: Emby.Server.Implementations/Data/SqliteUserRepository.cs (hunk 31 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/Data/SqliteUserRepository.cs]: Maybe for another day so we can get this one delivered, but why ever are we waiting for WRITE lock before executing a read only query!?

Comment (JustAMan) [reply Emby.Server.Implementations/Data/SqliteUserRepository.cs]: Can you move this code of getting the users inside `RetrieveAllUsers()` and make that function accept a `connection`? I don't like the amount of places where raw SQL is now...

Comment (Bond-009) [reply Emby.Server.Implementations/Data/SqliteUserRepository.cs]: > 
> 
> Maybe for another day so we can get this one delivered, but why ever are we waiting for WRITE lock before executing a read only query!?

We only have a write connection, 🤷‍♂️ maybe a bit of a misnomer.

**Comment by dkanada** [comment]
@Bond-009 @cvium it looks like this was almost done, but there aren't too many conflicts so we can merge this if it gets rebased against master and reviewed one last time.

File: Emby.Server.Implementations/Data/BaseSqliteRepository.cs [hunk 32]
```diff
@@ -199,40 +107,33 @@ protected bool TableExists(ManagedConnection connection, string name)
...
-        protected void RunDefaultInitialization(ManagedConnection db)
+        protected List<string> GetColumnNames(IDatabaseConnection connection, string table)
         {
-            var queries = new List<string>
-            {
-                "PRAGMA journal_mode=WAL",
-                "PRAGMA page_size=4096",
-                "PRAGMA synchronous=Normal"
-            };
+            var list = new List<string>();
```
Comment (cvium) [inline Emby.Server.Implementations/Data/BaseSqliteRepository.cs]: Not a very good variable name imo

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 33]
```diff
@@ -1909,24 +1901,21 @@ public List<ChapterInfo> GetChapters(BaseItem item)
         {
             CheckDisposed();
 
-            using (WriteLock.Read())
+            using (var connection = GetConnection(true))
             {
-                using (var connection = CreateConnection(true))
+                var list = new List<ChapterInfo>();
```
Comment (cvium) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: Variable name

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 34]
```diff
@@ -2624,67 +2604,64 @@ public List<BaseItem> GetItemList(InternalItemsQuery query)
                 }
             }
 
-            using (WriteLock.Read())
+            using (var connection = GetConnection(true))
             {
-                using (var connection = CreateConnection(true))
-                {
-                    var list = new List<BaseItem>();
+                var list = new List<BaseItem>();
```
Comment (cvium) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: Variable name

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 35]
```diff
@@ -2624,67 +2604,64 @@ public List<BaseItem> GetItemList(InternalItemsQuery query)
...
                     }
+                }
 
-                    // Hack for right now since we currently don't support filtering out these duplicates within a query
-                    if (query.EnableGroupByMetadataKey)
+                // Hack for right now since we currently don't support filtering out these duplicates within a query
+                if (query.EnableGroupByMetadataKey)
+                {
+                    var limit = query.Limit ?? int.MaxValue;
+                    limit -= 4;
```
Comment (cvium) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: Any idea what this is?

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs [hunk 36]
```diff
@@ -92,19 +92,15 @@ public class SqliteItemRepository : BaseSqliteRepository, IItemRepository
 
         private const string ChaptersTableName = "Chapters2";
 
-        protected override int? CacheSize => 20000;
```
Comment (cvium) [inline Emby.Server.Implementations/Data/SqliteItemRepository.cs]: Why was this removed?

**Comment by cvium** [comment]
While our use case is different, I still think this is an interesting read https://ericdraken.com/sqlite-performance-testing/

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs (hunk 35 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/SqliteItemRepository.cs]: No idea

File: Emby.Server.Implementations/Data/SqliteItemRepository.cs (hunk 36 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Data/SqliteItemRepository.cs]: Added it back
